File: apps/android/core/src/test/kotlin/com/safeworld/core/MatcherTest.kt
package com.safeworld.core

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/** Mirrors packages/core/test/matcher.test.ts — keep the two in sync. */
class MatcherTest {
    private fun lists(
        scam: List<String> = emptyList(),
        gambling: List<String> = listOf("bet365.com"),
        adult: List<String> = listOf("pornhub.com"),
    ): Map<CategoryId, Set<String>> = mapOf(
        CategoryId.SCAM to scam.toSet(),
        CategoryId.GAMBLING to gambling.toSet(),
        CategoryId.ADULT to adult.toSet(),
    )

    // region normalizeHost

    @Test
    fun `normalizeHost strips scheme, path, port, www and lowercases`() {
        assertEquals("bet365.com", Matcher.normalizeHost("HTTPS://WWW.Bet365.com:443/path?x=1"))
    }

    @Test
    fun `normalizeHost strips userinfo and trailing dot`() {
        assertEquals("sub.example.com", Matcher.normalizeHost("http://user:pass@sub.example.com./a"))
    }

    @Test
    fun `normalizeHost returns empty for blank input`() {
        assertEquals("", Matcher.normalizeHost("   "))
    }

    // endregion
    // region hostMatchesDomain

    @Test
    fun `hostMatchesDomain matches exact and subdomains but not siblings`() {
        assertTrue(Matcher.hostMatchesDomain("bet365.com", "bet365.com"))
        assertTrue(Matcher.hostMatchesDomain("m.bet365.com", "bet365.com"))
        assertFalse(Matcher.hostMatchesDomain("notbet365.com", "bet365.com"))
    }

    // endregion
    // region decide

    @Test
    fun `decide blocks a listed domain in an enabled category`() {
        val d = Matcher.decide("www.bet365.com", Settings.defaults(), lists())
        assertEquals(Matcher.BlockDecision(blocked = true, reason = "list2"), d)
    }

    @Test
    fun `decide blocks subdomains of a listed domain`() {
        assertTrue(Matcher.decide("sports.bet365.com", Settings.defaults(), lists()).blocked)
    }

    @Test
    fun `decide does not block when the master switch is off`() {
        val s = Settings.defaults().copy(enabled = false)
        assertFalse(Matcher.decide("bet365.com", s, lists()).blocked)
    }

    @Test
    fun `decide does not block when the category is disabled`() {
        val s = Settings.defaults().let {
            it.copy(categories = it.categories + (CategoryId.GAMBLING to false))
        }
        assertFalse(Matcher.decide("bet365.com", s, lists()).blocked)
    }

    @Test
    fun `decide custom allow overrides a category block`() {
        val s = Settings.defaults().copy(customAllow = listOf("bet365.com"))
        assertFalse(Matcher.decide("bet365.com", s, lists()).blocked)
    }

    @Test
    fun `decide custom block blocks an otherwise allowed domain`() {
        val s = Settings.defaults().copy(customBlock = listOf("example.com"))
        assertEquals(
            Matcher.BlockDecision(blocked = true, reason = "custom"),
            Matcher.decide("app.example.com", s, lists()),
        )
    }

    @Test
    fun `decide allows unlisted domains`() {
        assertFalse(Matcher.decide("wikipedia.org", Settings.defaults(), lists()).blocked)
    }

    // endregion
}

File: packages/core/test/matcher.test.ts
import { describe, it, expect } from "vitest";
import { normalizeHost, hostMatchesDomain, decide } from "../src/matcher.js";
import { defaultSettings } from "../src/settings.js";
import type { CategoryId } from "../src/categories.js";

function lists(over: Partial<Record<CategoryId, string[]>> = {}) {
  return {
    list1: new Set(over.list1 ?? []),
    list2: new Set(over.list2 ?? ["bet365.com"]),
    list3: new Set(over.list3 ?? ["pornhub.com"]),
  };
}

describe("normalizeHost", () => {
  it("strips scheme, path, port, www, and lowercases", () => {
    expect(normalizeHost("HTTPS://WWW.Bet365.com:443/path?x=1")).toBe("bet365.com");
  });
  it("strips userinfo and trailing dot", () => {
    expect(normalizeHost("http://user:pass@sub.example.com./a")).toBe("sub.example.com");
  });
  it("returns empty for blank input", () => {
    expect(normalizeHost("   ")).toBe("");
  });
});

describe("hostMatchesDomain", () => {
  it("matches exact and subdomains but not siblings", () => {
    expect(hostMatchesDomain("bet365.com", "bet365.com")).toBe(true);
    expect(hostMatchesDomain("m.bet365.com", "bet365.com")).toBe(true);
    expect(hostMatchesDomain("notbet365.com", "bet365.com")).toBe(false);
  });
});

describe("decide", () => {
  it("blocks a listed domain in an enabled category", () => {
    const d = decide("www.bet365.com", defaultSettings(), lists());
    expect(d).toEqual({ blocked: true, reason: "list2" });
  });

  it("blocks subdomains of a listed domain", () => {
    expect(decide("sports.bet365.com", defaultSettings(), lists()).blocked).toBe(true);
  });

  it("does not block when master switch is off", () => {
    const s = { ...defaultSettings(), enabled: false };
    expect(decide("bet365.com", s, lists()).blocked).toBe(false);
  });

  it("does not block when the category is disabled", () => {
    const s = defaultSettings();
    s.categories.list2 = false;
    expect(decide("bet365.com", s, lists()).blocked).toBe(false);
  });

  it("custom allow overrides a category block", () => {
    const s = { ...defaultSettings(), customAllow: ["bet365.com"] };
    expect(decide("bet365.com", s, lists()).blocked).toBe(false);
  });

  it("custom block blocks an otherwise-allowed domain", () => {
    const s = { ...defaultSettings(), customBlock: ["example.com"] };
    expect(decide("app.example.com", s, lists())).toEqual({ blocked: true, reason: "custom" });
  });

  it("allows unlisted domains", () => {
    expect(decide("wikipedia.org", defaultSettings(), lists()).blocked).toBe(false);
  });
});
